core/src/save/region/mod.rs

```diff
@@ -0,0 +1,834 @@
+        4096,
+    );
+
+    // Light
+    // convert raw lighting data (4bits / block) into a BitArray
+    let convert_light_data = |light_data: &Vec<i8>| {
+        let data = light_data
+            .chunks(8)
+            .map(|chunk| {
+                // not sure if there's a better (safe) way of doing this..
+                let chunk: [u8; 8] = [
+                    chunk[0] as u8,
+                    chunk[1] as u8,
+                    chunk[2] as u8,
+                    chunk[3] as u8,
+                    chunk[4] as u8,
+                    chunk[5] as u8,
+                    chunk[6] as u8,
+                    chunk[7] as u8,
+                ];
+                u64::from_le_bytes(chunk)
+            })
+            .collect();
+        BitArray::from_raw(data, 4, 4096)
+    };
+
+    if section.block_light.len() != 2048 || section.sky_light.len() != 2048 {
+        return Err(Error::IndexOutOfBounds);
+    }
+
+    let block_light = convert_light_data(&section.block_light);
+    let sky_light = convert_light_data(&section.sky_light);
+
+    let chunk_section = ChunkSection::new(data, Some(palette), block_light, sky_light);
+
+    if section.y >= 16 {
+        // Haha... nope.
+        return Err(Error::IndexOutOfBounds);
+    }
+
+    chunk.set_section_at(usize::from(section.y as u8), Some(chunk_section));
+
+    Ok(())
+}
+
+fn chunk_to_chunk_root(chunk: &Chunk, entities: Vec<EntityData>) -> ChunkRoot {
+    ChunkRoot {
+        level: ChunkLevel {
+            x_pos: chunk.position().x,
+            z_pos: chunk.position().z,
+            sections: chunk
+                .sections()
+                .iter()
+                .enumerate()
+                .filter_map(|(y, sec)| sec.map(|sec| (y, sec.clone())))
+                .map(|(y, section)| {
+                    LevelSection {
+                        y: y as i8,
+                        states: section.data().inner().iter().map(|x| *x as i64).collect(),
+                        palette: section
+                            .palette()
+                            .unwrap_or(&vec![])
+                            .iter()
+                            .map(|id| {
+                                let block = Block::from_native_state_id(*id).unwrap();
+
+                                let (name, props) = block.to_name_and_props();
+
+                                let mut prop_map = HashMap::new();
+                                props.into_iter().for_each(|(name, value)| {
+                                    prop_map.insert(name.to_string(), value);
+                                });
+
+                                LevelPaletteEntry {
+                                    name: name.to_string(),
+                                    props: Some(LevelProperties { props: prop_map }),
+                                }
+                            })
+                            .collect(), // TODO: what happens with the global palette?
```

Comment:
I could be wrong, but I don't think there is a "global palette" for local chunk saving.